.doc/s3-prerelease-images.md

```diff
@@ -0,0 +1,178 @@
+  e.g. `pr-123`.
+- In the `runtime-api` repository images are built on any feature branch.
+- Production SaaS releases are kicked off when a tag using semver format is
+  pushed to the OpenHands repo. Deployment of SaaS is delegated to workflows in
+  the `deploy` repo.
+- When a release includes changes to the `runtime-api` image, the `runtime-api`
+  repo needs to be tagged with the release version as well.
+
+## 3. Developer Experience
+
+### 3.1 Future State Process - Typical Releases
+
+S3 is now automatically updated when the image repos are on tagging of the
+release. No process change for the release engineer or developer.
+
+### 3.2 Future State Process - Troubleshooting Releases
+
+Engineers will go to a new Github Workflow in `OpenHands` and/or `runtime-api`
+to manually trigger a build supplying the target branch name. The image(s) are
+built and published to S3. An S3 link in the build log may be shared with the
+customer.
+
+## 4. Solution Design
+
+### 4.1 S3 Bucket Structure
+
+Images in the S3 bucket will be namespaced by branch name and commit SHA. If
+engineers use the same branch name across repos, images for a particular issue
+can be stored together. Since multiple builds could be produced from the same
+branch, the commit SHA is used within the branch folder to ensure uniqueness
+and traceability to the exact code being built.
+
+This approach aligns with the standard GHCR workflow which uses commit SHAs for
+image tagging, ensuring consistency across our build processes.
+
+To help clients identify the latest prerelease image from a feature branch, each
+branch folder will include a `CHANGELOG.md` file that documents all builds for
+that branch, with the most recent build listed first.
+
+Example:
+
+```plaintext
+s3://prerelease-bucket/
+├── jps/publish-prerelease-images/
+│   ├── CHANGELOG.md  # Documents all builds
+│   ├── abc1234/
+│   │   ├── openhands:latest.tar
+│   │   ├── openhands:main-abc1234.tar
+│   │   ├── enterprise:latest.tar
+│   │   ├── enterprise:main-abc1234.tar
+│   │   ├── runtime-api:latest.tar
+│   │   └── runtime-api:main-abc1234.tar
+│   └── def5678/
+│       ├── openhands:latest.tar
+│       ├── enterprise:latest.tar
+│       └── runtime-api:latest.tar
+└── mj/security-patch-123/
+    ├── CHANGELOG.md
+    └── ghi9012/
+        ├── openhands:hotfix-security.tar
+        └── enterprise:hotfix-security.tar
+```
+
+## 5. Open Questions
+
+### 5.1 Where does the push to S3 live?
+
+- Delegate to `deploy` repo GitHub action or do S3 push directly from OpenHands repo?
+
+### 5.2 Secrets
+
+- How do we secure AWS S3 secret? In OpenHands repo GitHub Actions Secrets or
+  via another mechanism?
+
+### 5.3 White List
+
+- How do we secure who can trigger these builds?
+
+### 5.4 Change Log
+
+Ideally this process should be self-documenting and leave a clear trail of what
+was released to the customer and why. The `CHANGELOG.md` file serves multiple purposes:
```

Comment:
The changelog part seems to be taking on an independent requirement. It's probably a good idea, but if it's non-trivial to make consider something like a `build.json` that just has info about the version and build time. Might want that anyway if some of the metadata should be machine readable.